File: apps/api/src/types/track.ts
import z from "zod";

// Replace the curly right single quotation mark (U+2019, `’`) with the ASCII
// apostrophe (U+0027, `'`) so artists like "Guns N’ Roses" and "Guns N' Roses"
// collapse to a single canonical form. Mirrors `normalize_text` in
// crates/mirror/src/track.rs.
const canonicalText = z
  .string()
  .nonempty()
  .trim()
  .transform((s) => s.replace(/’/g, "'"));

export const trackSchema = z.object({
  title: canonicalText,
  artist: canonicalText,
  artists: z
    .array(
      z.object({
        mbid: z.string().optional().nullable(),
        name: canonicalText,
      }),
    )
    .optional()
    .nullable(),
  album: canonicalText,
  albumArtist: canonicalText,
  duration: z.number(),
  mbId: z.string().optional().nullable(),
  isrc: z.string().optional().nullable(),
  albumArt: z.string().optional().nullable(),
  trackNumber: z.number().optional().nullable(),
  releaseDate: z
    .string()
    .regex(/^\d{4}-\d{2}-\d{2}$/, "Invalid date format. Use YYYY-MM-DD.")
    .transform((v) => new Date(v))
    .optional()
    .nullable(),

  year: z.number().optional().nullable(),
  discNumber: z.number().optional().nullable(),
  lyrics: z.string().optional().nullable(),
  composer: z.string().optional().nullable(),
  copyrightMessage: z.string().optional().nullable(),
  label: z.string().optional().nullable(),
  artistPicture: z.string().optional().nullable(),
  spotifyLink: z.string().optional().nullable(),
  lastfmLink: z.string().optional().nullable(),
  youtubeLink: z.string().optional().nullable(),
  tidalLink: z.string().optional().nullable(),
  appleMusicLink: z.string().optional().nullable(),
  deezerLink: z.string().optional().nullable(),
  timestamp: z.number().optional().nullable(),
  genres: z.array(z.string()).optional().nullable(),
});

export type Track = z.infer<typeof trackSchema>;

export type MusicbrainzTrack = {
  trackMBID: string;
  releaseMBID: string;
  name: string;
  artist: {
    id: string;
    mbid: string;
    name: string;
  }[];
  album: string;
  timestamp: string;
};

File: crates/mirror/src/track.rs
//! Normalized track payload posted to `app.rocksky.scrobble.createScrobble`.
//!
//! Matches the Zod schema at `apps/api/src/types/track.ts`. Only the four
//! required fields are mandatory; everything else is best-effort enrichment
//! from the source's response.

use chrono::{DateTime, Utc};
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct NormalizedTrack {
    pub title: String,
    pub artist: String,
    pub album: String,
    #[serde(rename = "albumArtist")]
    pub album_artist: String,

    /// Track duration in **milliseconds** (matches the Track schema).
    pub duration: i64,

    /// Unix seconds for when the play occurred (Track schema allows this).
    pub timestamp: i64,

    #[serde(skip_serializing_if = "Option::is_none", rename = "mbId")]
    pub mb_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub isrc: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "albumArt")]
    pub album_art: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "spotifyLink")]
    pub spotify_link: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "lastfmLink")]
    pub lastfm_link: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "trackNumber")]
    pub track_number: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "discNumber")]
    pub disc_number: Option<i32>,
}

impl NormalizedTrack {
    pub fn at(&self) -> DateTime<Utc> {
        DateTime::from_timestamp(self.timestamp, 0).unwrap_or_else(Utc::now)
    }
}

/// Trim and canonicalize a free-text field from an upstream provider.
///
/// Replaces the curly right single quotation mark (U+2019, `’`) with the
/// ASCII apostrophe (U+0027, `'`) so that artists like "Guns N’ Roses" and
/// "Guns N' Roses" collapse to a single canonical form before dedup and the
/// XRPC post.
pub fn normalize_text(s: &str) -> String {
    s.trim().replace('\u{2019}', "'")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalizes_curly_apostrophe() {
        assert_eq!(normalize_text(" Guns N\u{2019} Roses "), "Guns N' Roses");
        assert_eq!(normalize_text("Guns N' Roses"), "Guns N' Roses");
    }
}
